src/Core/Cart/Fees.php

```diff
@@ -64,6 +64,12 @@ class Fees
      */
     protected $isProcessed = false;
 
+    public function __construct()
+    {
+        $this->shippingFees = new AmountImmutable();
```

Comment:
When `$this->getFees()->getInitialShippingFees()` is called on `Calculator:: getDiscountTotal()` it's going to go through a huge number of legacy and complex function calls which is going to try to compute fees by getting carriers list, delivery options, carrier whatever-I-dont-know ... and at some point it's going to call `Order::getTotal()` which is going to call `Calculator::getDiscountTotal()`

At this moment `$this->getFees()->getInitialShippingFees()` will return `null` and next line will execute `$amount = $amount->sub(null);` which will crash.

From my debugging process it seems that when not computed yet `$this->shippingFees` should be an `AmountImmutable` instance with properties set  to `0.0`. So I initialize this property here in order for it to never ne null, and I hope it works 🙏 .

According to our behat tests it works 🤔

Reply:
Finally I found a comment which describes me in front of Calculator and CartRuleCalculator files . This re-comfort me now, I was thinking I am the only one who doesn't understand coding patterns :D

Reply:
@arouiadib Unfortunately no you are not insane. It's this code which is insane. Recursion loop happen in a lot of computations for product prices, cart total prices and cart rules prices and is making any attempt to fix part of the code an herculean task.

Unfortunately we cannot delete all this code because of SemVer althought it seems to be the only relevant solution to fix all these crazy cart rules bugs. It must wait for PS 8.0.0 where we will be allowed to perform such a huge BC break.

We had the same issue for PR https://github.com/PrestaShop/PrestaShop/pull/12878

Reply:
@matks , thank you for your usual informative comments. I totally agree with you in parts I understand.

I think one of the good points is that there is a number of behat tests now. Till the day comes, we will be adding more tests, and no matter is the complexity of the current system, the swap to a new system will be easier in 1.8. I hope I am not wrong!
For sure, you are not meaning PS 8 but PS 1.8 :p :)
One another thing I see about cart rules, is that, still we need to discuss them before fixing any issue (in current system or in 1.8). If we take the current bug you are fixing as example, some merchants may think that free shipping should be cumulative and that free shipping vouchers should be stack in total (we can discuss this point later if you want but it is not the purpose of my comment). 

The problem is when that merchant raises a ticket here in github, that might end up with a lot of time wasted if there is no link to send back telling him how Prestashop is supposed to calculate his cart rule. In other words, what is the equation Prestashop will be using to calculate that cart rule, for all times.

He can decide then whether to wait for the fix from github or to hire a developer. For Prestashop, it is not too much questions to POs and too much rethininking of things, and...and... all the advantages of specs in software industry.

I still believe that cart rules specs must be the first written in the whole specs of Prestashop due to its complexity, relativity with total amount and invoices....

Sorry for being long

Reply:
What I wrote is not maybe in the appropriate time and tone. So don't take this matter seriously :redface

Reply:
> @matks , thank you for your usual informative comments. I totally agree with you in parts I understand.
> 
> I think one of the good points is that there is a number of behat tests now. Till the day comes, we will be adding more tests, and no matter is the complexity of the current system, the swap to a new system will be easier in 1.8. I hope I am not wrong!
> For sure, you are not meaning PS 8 but PS 1.8 :p :)

I mean PS 8.0.0 .
Although it is not decided 100% yet we think that, for next major version, we will stop using the 1.X.Y.Z version numbers and start using the X.Y.Z version number standard.

So if PS 1.7.9 is the last PS 1.7 minor release, next release will be PS 8.0.0 .
It changes nothing about the project, it's just a version number. PS 1.8.0.0 or PS 8.0.0 is the same.

> One another thing I see about cart rules, is that, still we need to discuss them before fixing any issue (in current system or in 1.8). If we take the current bug you are fixing as example, some merchants may think that free shipping should be cumulative and that free shipping vouchers should be stack in total (we can discuss this point later if you want but it is not the purpose of my comment).

@Robin-Fischer-PS (who is the author of the issue which states free shipping vouchers should not be cumulative) is always in discussion with Product Team and Product Team is writing down these specs into https://github.com/PrestaShop/prestashop-specs

> The problem is when that merchant raises a ticket here in github, that might end up with a lot of time wasted if there is no link to send back telling him how Prestashop is supposed to calculate his cart rule. In other words, what is the equation Prestashop will be using to calculate that cart rule, for all times.

Indeed it must go into https://github.com/PrestaShop/prestashop-specs . Product Team is on this topic

> Sorry for being long

I actually appreciate you took so much time to write this comment, it means you care about the project and we ❤️ having dedicated contributors

> What I wrote is not maybe in the appropriate time and tone. So don't take this matter seriously 😊

Your tone and time are appropriate and the topic you raise is very important. If you have other similar topics / feedbacks please share them with us, it's very important and we appreciate this. Your feedbacks have a lot of value 👍

Reply:
I thank you very much @matks . 
It is always a great pleasure to be part of a big masonry. It is a beautiful building, where just one brick contribution makes you feel helpful. As you know, Prestashop is a great source of income in normal days, and a rescue for many businesses in special days. For me, it a also special building: a university where I always learn new things from experts.

I will be glad to contribute in "free Shipping discount" specs. I will be writing comments and sharing them with @Robin-Fischer-PS so that we can think of all possible scenarios.

Again thank you all guys :)

Reply:
Hello @Robin-Fischer-PS ,
I have gathered what I think about the shipping discount calculation in: https://github.com/PrestaShop/prestashop-specs/issues/107. I will be glad to explain more if things are unclear.
Regards